core/src/main/java/com/linecorp/armeria/internal/common/DefaultSplitHttpResponse.java

```diff
@@ -0,0 +1,335 @@
+    }
+
+    @Override
+    public void abort() {
+        response.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        response.abort(cause);
+    }
+
+    private final class BodySubscriber implements Subscriber<HttpObject>, Subscription {
+
+        @Nullable
+        private Throwable cause;
+
+        private boolean completing;
+        // 1 is used for prefetching headers
+        private long pendingRequests = 1;
+
+        @Nullable
+        volatile Subscriber<? super HttpData> downstream;
+        @Nullable
+        private volatile Subscription upstream;
+
+        private volatile boolean cancelCalled;
+
+        private void setDownStream(Subscriber<? super HttpData> downstream) {
+            try {
+                if (!downstreamUpdater.compareAndSet(this, null, downstream)) {
+                    downstream.onSubscribe(NoopSubscription.get());
+                    downstream.onError(new IllegalStateException("subscribed by other subscriber already"));
+                    return;
+                }
+                downstream.onSubscribe(this);
+                if (cause != null) {
+                    downstream.onError(cause);
+                } else if (completing) {
+                    downstream.onComplete();
+                }
+            } catch (Throwable t) {
+                throwIfFatal(t);
+                logger.warn("Subscriber should not throw an exception. subscriber: {}", downstream, t);
+            }
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (cancelCalled || upstream != null) {
+                subscription.cancel();
+                return;
+            }
+            upstream = subscription;
+            subscription.request(pendingRequests);
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                // Just abort the publisher so subscriber().onError(e) is called and resources are cleaned up.
+                response.abort(new IllegalArgumentException(
+                        "n: " + n + " (expected: > 0, see Reactive Streams specification rule 3.9)"));
+                return;
+            }
+            if (executor.inEventLoop()) {
+                request0(n);
+            } else {
+                executor.execute(() -> request0(n));
+            }
+        }
+
+        private void request0(long n) {
+            final Subscription upstream = this.upstream;
+            if (upstream == null) {
+                pendingRequests = LongMath.saturatedAdd(n, pendingRequests);
+            } else {
+                upstream.request(n);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            if (cancelCalled) {
+                return;
+            }
+            cancelCalled = true;
+            downstream = NoopSubscriber.get();
```

Comment:
If subscribed with `SubscriptionOptions.NOTIFY_CANCELLATION`, the subscriber doesn't get notified because of this logic? Could you add a test for this?

Reply:
That's a good point. I've add test code to pass Reactive Streams TCK. Let me update this a lit. :-)

Reply:
question: Can't we just leave the `downstream`? (i.e. not assign the `NoopSubscriber`?)

Reply:
AFAIK, if `Subscription.cancel()` called, the publisher should drop the subscriber's reference.
https://github.com/reactive-streams/reactive-streams-jvm#3.13
If not, `required_spec313_cancelMustMakeThePublisherEventuallyDropAllReferencesToTheSubscriber` in TCK failed.

Reply:
Thanks for the explanation. 😄